### ISSUE TITLE
[ufoLib.glifLib] open path validation forbids trailing off curve point

### ISSUE THREAD

**Issue by justvanrossum** [fonttools/fonttools#2972]
https://github.com/fonttools/fonttools/blob/2f1f5e5e7be331d960a0e30d537c2b4c70d89285/Lib/fontTools/ufoLib/glifLib.py#L1900-L1903

The UFO spec doesn't forbid such a thing, and it is useful during the drawing process: while constructing a shape a traling off curve point is very common, and being able to save the glyph in that state should be possible.

Fontra works around this by not validating .glif writes, which is not ideal:

https://github.com/googlefonts/fontra/blob/dfe0cbc827a616b053c1f6da044a109652e5a471/src/fontra/backends/designspace.py#L222-L224

I'll gladly make a PR removing this check, if we can agree that it is indeed bogus.

**Comment by justvanrossum** [comment]
The check was introduced here: https://github.com/fonttools/fonttools/commit/69a04a7d7a3f026a80988ce25797ed70237d1371

@typesupply, do you remember why this was needed at the time?

**Comment by justvanrossum** [comment]
The commit before quietly filters the trailing off curve points: https://github.com/fonttools/fonttools/commit/64ebada787531e84f7842ddf530cc24c7c3ea37d, but also doesn't say why.

**Comment by anthrotype** [comment]
perhaps because one wouldn't know how to even draw a curve segment without the oncurve point at the end. The trailing off-curves basically float in the void...

**Comment by justvanrossum** [comment]
For a path it makes no sense (and should be ignored), but for a font editor, the node/handle can be displayed.

**Comment by anthrotype** [comment]
yeah, I see you what you mean. I'd be OK to remove the check.

**Comment by benkiel** [comment]
Not that I want to make the validation more complicated, but maybe there could be a validation set for "messy drawings" and one for "I expect to make a working font from this UFO, what is wrong with it?"

**Comment by justvanrossum** [comment]
Aren't open paths by themselves already firmly in the domain of "messy drawings"?

**Comment by benkiel** [comment]
Right, they are; I’m spitballing adding a validation mode that would allow open paths. Maybe “design” vs “production”. This may (likely) be a not great idea, but having a ufo that passes validation but would break font generation also seems not great.

**Comment by anthrotype** [comment]
> a ufo that passes validation but would break font generation

i'm actually not sure if having trailing off curve points in an open path would specifically break font generation. I think the _open_ paths themselves are the problematic part (I believe booleanOperations throws an error if a path is not closed, and even in absence of overlap removal, the ttGlyphPen errors out with open paths IIRC).

**Comment by typesupply** [comment]
> @typesupply, do you remember why this was needed at the time?

I don't remember writing this code in detail. I remember thinking through this and discussing teh font editor use case you mentioned with you or Frederik or Erik or someone on iChat or Slack or whatever we were using then. Looking at the code I can reconstruct what I may have been thinking. (Most of what follows is obvious, but I'm going to write it out for clarity.)

Let's say I'm in a GUI editor using an Illustrator-like pen tool to draw a circle. I draw my first segment and stop drawing because the phone rings and I need to answer it. I instinctively hit command-s and turn my attention to the phone call. The way the pen tools commonly work is that they preemptively create the next offcurve because that's probably what the user wants to do. Here's an example I just drew in RoboFont using the pen tool:

<img width="559" alt="image" src="https://user-images.githubusercontent.com/4992934/215630396-f5e731c2-7d26-44b6-b8e7-4d83f3d43c3f.png">

```xml
move (0, 0)
off curve (0, 60)
off curve (40, 100)
curve (100, 100)
off curve (160, 100)
```

When I save the file, should that off curve be written into the GLIF? If so, how should it be interpreted by a tool *other than the one that created it*? In other words, it would make sense to the user if that off curve was visible in a subsequent RoboFont session picking up where they left off. But what is another tool supposed to think that loose off curve means? It doesn't belong to a curve so it's just hanging out there as an artifact that only makes sense to one particular tool.

I probably only saw ambiguity in the interpretation, classified this as a "tool problem" and maybe didn't feel comfortable changing the spec, so I put the block in place. I have no opinion on if this was the right decision or not.

> Aren't open paths by themselves already firmly in the domain of "messy drawings"?

One tiny use case where open paths are very important. [I used open paths while drawing this.](https://typesupply.com/fonts/ohm) I still have the UFO sources with the central paths that were used to create the final tubes.

**Comment by justvanrossum** [comment]
> how should it be interpreted by a tool _other than the one that created it_? 

Hm, I don't quite follow this argument. In my eyes it is perfectly obvious what dangling off curve points means: an unfinished continuation of the path. You can complete it by adding an on-curve point. Or ignore it if you need segments, which is what PointToSegmentPen does.

Another issue is symmetry: while UFO is capable of storing trailing off-curve points, it cannot store leading off-curve points, because an open path must start with a "move", which is on-curve. I consider this a design flaw of UFO (that mostly likely I am responsible for). The "closed" or "open" property of the path should have been a separate property of the contour, and not implied by the first point type. Anyway, I digress.

I'm happy to drop this issue — I don't think the use case of *storing* an incomplete path is all that compelling after all (esp. given the asymmetry mentioned above).

My main point stands, though: the validator should ideally not forbid something that is legal according to the spec.

Btw. your "tiny use case" is a fair point: I use it like that all the time, too.

**Comment by typesupply** [comment]
> Hm, I don't quite follow this argument. In my eyes it is perfectly obvious what dangling off curve points means: an unfinished continuation of the path. You can complete it by adding an on-curve point. Or ignore it if you need segments, which is what PointToSegmentPen does.

Right. I really have no opinion on what should be done. I was trying to figure out my logic. Another thing I could have been thinking about: "Will this situation break any existing pens, modules, etc.?" I was simultaneously writing the UFO 3 spec, updating ufoLib, updating defcon and reviewing all of the pens/point pens and I spent a lot of time worrying about things breaking.

> My main point stands, though: the validator should ideally not forbid something that is legal according to the spec.

I agree. Either the spec or the validator need to change.
